nexus/src/app/sagas/instance_delete.rs

```diff
@@ -67,6 +71,79 @@ impl NexusSaga for SagaInstanceDelete {
 
 // instance delete saga: action implementations
 
+async fn sid_delete_network_config(
+    sagactx: NexusActionContext,
+) -> Result<(), ActionError> {
+    let params = sagactx.saga_params::<Params>()?;
+    let opctx = crate::context::op_context_for_saga_action(
+        &sagactx,
+        &params.serialized_authn,
+    );
+    let osagactx = sagactx.user_data();
+    let dpd_client = &osagactx.nexus().dpd_client;
+    let datastore = &osagactx.datastore();
+    let log = sagactx.user_data().log();
+
+    debug!(log, "fetching external ip addresses");
+
+    let external_ips = &datastore
+        .instance_lookup_external_ips(&opctx, params.authz_instance.id())
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    // TODO: currently if we have this environment variable set, we want to
+    // bypass all calls to DPD. This is mainly to facilitate some tests where
+    // we don't have dpd running. In the future we should probably have these
+    // testing environments running dpd-stub so that the full path can be tested.
+    if let Ok(_) = std::env::var("SKIP_ASIC_CONFIG") {
+        return Ok(());
+    };
+
+    let mut errors: Vec<ActionError> = vec![];
+
+    // Here we are attempting to delete every existing NAT entry while deferring
+    // any error handling. If we don't defer error handling, we might end up
+    // bailing out before we've attempted deletion of all entries.
+    for entry in external_ips {
+        debug!(log, "deleting nat mapping for entry: {entry:#?}");
+        let result = match entry.ip {
+            ipnetwork::IpNetwork::V4(network) => {
+                dpd_client
+                    .nat_ipv4_delete(&network.ip(), *entry.first_port)
+                    .await
+            }
+            ipnetwork::IpNetwork::V6(network) => {
+                dpd_client
+                    .nat_ipv6_delete(&network.ip(), *entry.first_port)
+                    .await
+            }
+        };
+
+        match result {
+            Ok(_) => {
+                debug!(log, "deletion of nat entry successful for: {entry:#?}");
+            }
+            Err(e) => {
+                if e.status() == Some(http::StatusCode::NOT_FOUND) {
+                    debug!(log, "no nat entry found for: {entry:#?}");
+                } else {
+                    let new_error =
+                        ActionError::action_failed(Error::internal_error(
+                            &format!("failed to delete nat entry via dpd: {e}"),
+                        ));
+                    errors.push(new_error);
+                }
+            }
+        }
+    }
+
+    if let Some(error) = errors.first() {
+        return Err(error.clone());
+    }
```

Comment:
I think returning the first error is probably the best we can do at this point, but it might help to log all of them?

Reply:
Added logging above.